skimage/morphology/_util.py

```diff
@@ -51,35 +54,206 @@ def _validate_connectivity(image_dim, connectivity, offset):
+        `selem` is not None.
+    ndim : int
+        Number of dimensions `selem` ought to have.
+
+    Returns
+    -------
+    selem : ndarray
+        Validated or new structuring element specifying the neighborhood.
+
+    Examples
+    --------
+    >>> _resolve_neighborhood(None, 1, 2)
+    array([[False,  True, False],
+           [ True,  True,  True],
+           [False,  True, False]])
+    >>> _resolve_neighborhood(None, None, 3).shape
+    (3, 3, 3)
     """
-    structure = structure.copy()  # Don't modify original input
-    structure[tuple(center)] = 0  # Ignore the center; it's not a neighbor
-    connection_indices = np.transpose(np.nonzero(structure))
-    offsets = (np.ravel_multi_index(connection_indices.T, image_shape,
-                                    order=order) -
-               np.ravel_multi_index(center, image_shape, order=order))
-    squared_distances = np.sum((connection_indices - center) ** 2, axis=1)
-    return offsets[np.argsort(squared_distances)]
+    if selem is None:
+        if connectivity is None:
+            connectivity = ndim
+        selem = ndi.generate_binary_structure(ndim, connectivity)
+    else:
+        # Validate custom structured element
+        selem = np.asarray(selem, dtype=np.bool)
+        # Must specify neighbors for all dimensions
+        if selem.ndim != ndim:
+            raise ValueError(
+                "structuring element and image must have the same number of "
+                "dimensions"
+            )
+        # Must only specify direct neighbors
+        if any(s != 3 for s in selem.shape):
+            raise ValueError("dimension size in structuring element is not 3")
+
+    return selem
+
+
+def _set_edge_values_inplace(image, value):
+    """Set edge values along all axes to a constant value.
+
+    Parameters
+    ----------
+    image : ndarray
+        The array to modify inplace.
+    value : scalar
+        The value to use. Should be compatible with `image`'s dtype.
+
+    Examples
+    --------
+    >>> image = np.zeros((4, 5), dtype=int)
+    >>> _set_edge_values_inplace(image, 1)
+    >>> image
+    array([[1, 1, 1, 1, 1],
+           [1, 0, 0, 0, 1],
+           [1, 0, 0, 0, 1],
+           [1, 1, 1, 1, 1]])
+    """
+    for axis in range(image.ndim):
+        # Index first and last element in each dimension
+        sl = (slice(None),) * axis + ((0, -1),) + (...,)
+        image[sl] = value
+
+
+def _fast_pad(image, value):
+    """Pad an array on all axes by one with a value.
+
+    Parameters
+    ----------
+    image : ndarray
+        Image to pad.
+    value : scalar
+         The value to use. Should be compatible with `image`'s dtype.
+
+    Returns
+    -------
+    padded_image : ndarray
+        The new image.
+
+    Notes
+    -----
+    The output of this function is equivalent to::
+
+        np.pad(image, 1, mode="constant", constant_values=value)
+
+    Up to versions < 1.17 `numpy.pad` uses concatenation to create padded
```

Comment:
Add this to TODO.txt? Otherwise we'll forget about it when we bump to numpy >= 1.17.

Reply:
Definitely!